extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java

```diff
@@ -41,19 +45,30 @@ private void doHandle(RoutingContext ctx) {
             Arc.container().instance(CurrentIdentityAssociation.class).get().setIdentity(user.getSecurityIdentity());
         }
         SmallRyeHealthReporter reporter = Arc.container().instance(SmallRyeHealthReporter.class).get();
-        SmallRyeHealth health = getHealth(reporter, ctx);
-        HttpServerResponse resp = ctx.response();
-        if (health.isDown()) {
-            resp.setStatusCode(503);
-        }
-        resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
-        Buffer buffer = Buffer.buffer(256); // this size seems to cover the basic health checks
-        try (BufferOutputStream outputStream = new BufferOutputStream(buffer);) {
-            reporter.reportHealth(outputStream, health);
-            resp.end(buffer);
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
+        getHealth(reporter, ctx).subscribe().with(health -> {
+            HttpServerResponse resp = ctx.response();
+            if (health.isDown()) {
+                resp.setStatusCode(503);
+            }
+            resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
+            Buffer buffer = Buffer.buffer(256); // this size seems to cover the basic health checks
+            try (BufferOutputStream outputStream = new BufferOutputStream(buffer)) {
+                reporter.reportHealth(outputStream, health);
```

Comment:
It's rather annoying that it uses an output stream which is considered blocking by nature. Is there anything we can do to avoid this?

Reply:
The output stream used here (and also in the original code) is writing to an in-memory buffer, and is therefore non-blocking. 

When that memory buffer is written to the response using `end(buffer)`, this is also non-blocking as this returns a `Future`. 

The reason there is explicit error handling here are the checked exceptions of the `OutputStream`, and the potential case when `reportHealth()` throws an exception. As stated above, I verified that any exception thrown in a health check is already handled / wrapped in SmallRye, so those exceptions won't show up here.

Reply:
To add to that: `reportHealth()` is only about encoding the `health` status to JSON, all checks have already been executed at that point.

Reply:
@cescoffier - `reporter.reportHealth()` is part of SmallRye health and it is outside of the Quarkus project. It would require a PR to that library and a release of that library first. 

Given that `reportHealth()` uses Jakarta to serialize a pre-computed `JsonObject`, I would hope that this PR is good enough as it is now to be merged.